src/solvers/refinement/string_refinement.cpp

```diff
@@ -1320,112 +1380,129 @@ exprt fill_in_array_expr(const array_exprt &expr, std::size_t string_max_length)
   return result;
 }
 
-/// create an equivalent expression where array accesses and 'with' expressions
+/// Create an equivalent expression where array accesses are replaced by 'if'
+/// expressions: for instance in array access `arr[index]`, where:
+/// `arr := {12, 24, 48}` the constructed expression will be:
+///    `index==0 ? 12 : index==1 ? 24 : 48`
+static exprt substitute_array_access(
+  const array_exprt &array_expr,
+  const exprt &index,
+  const std::function<symbol_exprt(const irep_idt &, const typet &)>
+    &symbol_generator)
+{
+  const typet &char_type = array_expr.type().subtype();
+  const std::vector<exprt> &operands = array_expr.operands();
+
+  exprt result = symbol_generator("out_of_bound_access", char_type);
+
+  for(std::size_t i = 0; i < operands.size(); ++i)
+  {
+    // Go in reverse order so that smaller indexes appear first in the result
+    const std::size_t pos = operands.size() - 1 - i;
```

Comment:
Why not iterate backwards instead of defining a `pos` variable?

Reply:
`pos` is used in the `equal_exprt` as an integer, so we cannot use iterators.

Reply:
But what about `for(std::size_t pos = operands.size() - 1; pos >= 0; --pos)`? Oh, is `size_t` unsigned?

Reply:
testing for a size_t being greater than 0 will always return true